File: gpstracker/src/main/java/com/golriz/gpstracker/LocationTrackerService.java
package com.golriz.gpstracker;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.location.Location;
import android.location.LocationListener;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;
import androidx.annotation.Nullable;

import java.util.Timer;
import java.util.TimerTask;

import static android.content.ContentValues.TAG;

public class LocationTrackerService extends Service implements LocationListener {
    public int counter = 0;
    Context context;
    public static boolean stopTime = false;
    private int newLocationDistance = 0;
    private int newLocationTimeInterval = 0;
    private int syncToserverTimeInterval = 0;
    private int numberOfRecordsToSync = 0;

    public LocationTrackerService(Context applicationContext, int newLocationDistance, int newLocationTimeInterval, int syncToserverTimeInterval, int numberOfRecordsToSync) {
        super();
        context = applicationContext;
        this.newLocationDistance = newLocationDistance;
        this.newLocationTimeInterval = newLocationTimeInterval;
        this.syncToserverTimeInterval = syncToserverTimeInterval;
        this.numberOfRecordsToSync = numberOfRecordsToSync;

    }

    public LocationTrackerService(Context context) {
        this.context = context;
    }

    public LocationTrackerService() {
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        super.onStartCommand(intent, flags, startId);
        startTimer();
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        Log.i("EXIT", "ondestroy!");

        Intent broadcastIntent = new Intent("service.locationTracker.stopped");
        sendBroadcast(broadcastIntent);
        stoptimertask();

    }

    private Timer timer;
    private TimerTask timerTask;

    public void startTimer() {
        Log.d(TAG, "startTimer: ");
        //set a new Timer
        timer = new Timer();

        //initialize the TimerTask's job
        initializeTimerTask();

        //schedule the timer, to wake up every 1 second
        timer.schedule(timerTask, 1000, 1000); //
    }

    public void initializeTimerTask() {
        timerTask = new TimerTask() {
            public void run() {
                Log.i("in timer", "in timer ++++  " + (counter++));
            }
        };
    }

    public void stoptimertask() {
        Log.d(TAG, "stoptimertask: ");

        //stop the timer, if it's not already null
        if (timer != null) {
            timer.cancel();
            timer = null;
        }
    }


    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onLocationChanged(Location location) {

    }

    @Override
    public void onStatusChanged(String s, int i, Bundle bundle) {

    }

    @Override
    public void onProviderEnabled(String s) {

    }

    @Override
    public void onProviderDisabled(String s) {

    }
}

File: gpstracker/src/main/java/com/golriz/gpstracker/Core/LocationTrackerService.kt
package com.golriz.gpspointer.Config

import android.app.Service
import android.content.Context
import android.content.Intent
import android.location.Location
import android.location.LocationListener
import android.os.Bundle
import android.os.IBinder
import android.util.Log

import java.util.Timer
import java.util.TimerTask

import android.content.ContentValues.TAG

class LocationTrackerService : Service, LocationListener {
    var counter = 0
    internal lateinit var context: Context
    private var newLocationDistance = 0
    private var newLocationTimeInterval = 0
    private var syncToserverTimeInterval = 0
    private var numberOfRecordsToSync = 0

    private var timer: Timer? = null
    private var timerTask: TimerTask? = null

    constructor(
        applicationContext: Context,
        newLocationDistance: Int,
        newLocationTimeInterval: Int,
        syncToserverTimeInterval: Int,
        numberOfRecordsToSync: Int
    ) : super() {
        context = applicationContext
        this.newLocationDistance = newLocationDistance
        this.newLocationTimeInterval = newLocationTimeInterval
        this.syncToserverTimeInterval = syncToserverTimeInterval
        this.numberOfRecordsToSync = numberOfRecordsToSync

    }

    constructor(context: Context) {
        this.context = context
    }

    constructor() {}

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        super.onStartCommand(intent, flags, startId)
        startTimer()
        return Service.START_STICKY
    }

    override fun onDestroy() {
        super.onDestroy()

        Log.i("EXIT", "ondestroy!")

        val broadcastIntent = Intent("service.locationTracker.stopped")
        sendBroadcast(broadcastIntent)
        stoptimertask()

    }

    fun startTimer() {
        Log.d(TAG, "startTimer: ")
        //set a new Timer
        timer = Timer()

        //initialize the TimerTask's job
        initializeTimerTask()

        //schedule the timer, to wake up every 1 second
        timer!!.schedule(timerTask, 1000, 1000) //
    }

    fun initializeTimerTask() {
        timerTask = object : TimerTask() {
            override fun run() {
                Log.i("in timer", "in timer ++++  " + counter++)
            }
        }
    }

    fun stoptimertask() {
        Log.d(TAG, "stoptimertask: ")

        //stop the timer, if it's not already null
        if (timer != null) {
            timer!!.cancel()
            timer = null
        }
    }


    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onLocationChanged(location: Location) {

    }

    override fun onStatusChanged(s: String, i: Int, bundle: Bundle) {

    }

    override fun onProviderEnabled(s: String) {

    }

    override fun onProviderDisabled(s: String) {

    }

    companion object {
        var stopTime = false
    }
}